{"metadata":{"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"name":"python","version":"3.10.11","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat_minor":5,"nbformat":4,"cells":[{"cell_type":"markdown","source":"This file is a Jupyter notebook. It contains text (as here), written in the language Markdown and bits of Python code. Text and code are in cells. Content of cells can be executed directly in your browser ('Run' button above).\nYou can add, remove, move cells - when they are in the 'blue' mode - and edit them, when they are in the 'green' mode. To change modes, (double) click on cells. \nAfter the cell content was executed, you can edit it again.","metadata":{},"id":"29ab0b06"},{"cell_type":"markdown","source":"This cell is also in Markdown. Check what happens when you change its format to 'Code' in the drop-down option above.","metadata":{},"id":"b341c63d"},{"cell_type":"markdown","source":"For the cells with code, Notebook will execute the code - assign variables, compute, plot etc","metadata":{},"id":"84bd0f97"},{"cell_type":"code","source":"4*3","metadata":{},"execution_count":null,"outputs":[],"id":"3859d016"},{"cell_type":"markdown","source":"#### Variables","metadata":{},"id":"6846d8b6"},{"cell_type":"code","source":"#This is a cell with code. Lines starting with # are treated as comments and not executed. Everything else will be interpreted as Python\n#Here we assign a number (13) to a variable number_of_animals_with_4_legs\nnumber_of_animals_with_4_legs = 13","metadata":{},"execution_count":null,"outputs":[],"id":"8e2d9f72"},{"cell_type":"code","source":"number_of_animals_with_4_legs","metadata":{},"execution_count":null,"outputs":[],"id":"f5b728d9"},{"cell_type":"code","source":"print(number_of_animals_with_4_legs)","metadata":{},"execution_count":null,"outputs":[],"id":"11b04bb7"},{"cell_type":"code","source":"number_of_legs_in_the_zoo = number_of_animals_with_4_legs","metadata":{},"execution_count":null,"outputs":[],"id":"bd475926"},{"cell_type":"markdown","source":"(Add a code cell below and) compute number of legs in a Zoo with 125 four-legged animals","metadata":{},"id":"9d393f43"},{"cell_type":"markdown","source":"What if you want to include legs from birds? Create a variable number_of_animals_with_2_legs and update the formula for the number of legs in the ZOO.\n","metadata":{},"id":"f93df23e"},{"cell_type":"markdown","source":"#### Types of variables","metadata":{},"id":"b32f29be"},{"cell_type":"code","source":"#Variables may be numbers (14, -8, 3.245)\na = 8\n#or strings ('Word','or a sentence')\nname = \"Taylor\"\n#or logical values (True, False)\nshe_sings_well = True","metadata":{},"execution_count":null,"outputs":[],"id":"42baaa5a"},{"cell_type":"markdown","source":"Create now variables of each type. Give them meaningful names - also add comments.\n\n\n\n","metadata":{},"id":"4946a5b8"},{"cell_type":"markdown","source":"#### Operations on variables","metadata":{},"id":"ccc1a317"},{"cell_type":"markdown","source":"What can be applied to a variable depends on its type. Numerical one may be trated with usual arithmetics, also compared with other variable (the result will be logical)","metadata":{},"id":"582a7a5f"},{"cell_type":"code","source":"#Compare number_of_animals_with_2_legs with number_of_animals_with_4_legs","metadata":{},"execution_count":26,"outputs":[],"id":"1ab792f7"},{"cell_type":"markdown","source":"You can  concatenate strings with '+'. Create a variable 'last_name', with value 'Swift' and use it together with the variable 'name'","metadata":{},"id":"52b4c8f6"},{"cell_type":"markdown","source":"### Operations on strings","metadata":{},"id":"848b4203"},{"cell_type":"code","source":"\n# String variables and literals\nprotein = \"GFP\"                    # Winner of 2008 Nobel in chemistry\nprotein_seq_begin = 'MSKGEELFTG'\nprotein_seq_end = 'HGMDELYK'\n\n ","metadata":{},"execution_count":null,"outputs":[],"id":"fd14a7d5"},{"cell_type":"code","source":" \n# Concatenation of strings\nprotein_seq = protein_seq_begin + '...' + protein_seq_end\nprint('Protein sequence of GFP: ' + protein_seq)\n","metadata":{},"execution_count":null,"outputs":[],"id":"6bd12dca"},{"cell_type":"code","source":" \n# String method str.upper()\nDNA_seq = 'atgagtaaag...actatacaaa'\nDNA_seq = DNA_seq.upper()\nprint('DNA sequence: ' + DNA_seq)\n \n","metadata":{},"execution_count":null,"outputs":[],"id":"915d32e4"},{"cell_type":"code","source":"# Forward index starts with 0 and increases\n# Backward index starts with -1 and decreases\nprint('The second nucleotide:', DNA_seq[1])\nprint('The last nucleotide:', DNA_seq[-1])\n","metadata":{},"execution_count":null,"outputs":[],"id":"f73bf70e"},{"cell_type":"code","source":" \n# Slicing a string\nfirst_codon = DNA_seq[0:3]         # index 3 excluded\nlast_codon = DNA_seq[-3:]\nprint('First codon:', first_codon)\nprint('Last codon:', last_codon)","metadata":{},"execution_count":null,"outputs":[],"id":"788b0bb7"},{"cell_type":"code","source":"'ATG' in DNA_seq\nDNA_seq.find('taaa'.upper())","metadata":{},"execution_count":null,"outputs":[],"id":"dd100a94"},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[],"id":"c1be18a4"},{"cell_type":"raw","source":"","metadata":{},"id":"f743d436"},{"cell_type":"raw","source":"","metadata":{},"id":"9285c9d4"},{"cell_type":"raw","source":"","metadata":{},"id":"926134c0"},{"cell_type":"markdown","source":"Now your turn! We have two CRISPR guides. Identify which one  targets  the CTLA4 sequence.","metadata":{},"id":"ab693be5"},{"cell_type":"code","source":"target_of_guide1= 'CAAGTGAACCTCACTATCCA'\ntarget_of_guide2= 'TGGGATCAAAGCTATCTATA' ","metadata":{"trusted":true,"tags":[]},"execution_count":2,"outputs":[],"id":"241cd98b"},{"cell_type":"code","source":"DNA_CTLA4= 'GCTTTCTATTCAAGTGCCTTCTGTGTGTGCACATGTGTAATACATATCTGGGATCAAAGCTATCTATATAAAGTCCTTGATTCTGTGTGGGTTCAAACACATTTCAAAGCTTCAGGATCCTGAAAGGTTTTGCTCTACTTCCTGAAGACCTGAACACCGCTCCCATAAAGCCATGGCTTGCCTTGGATTTCAGCGGCACAAGGCTCAGCTGAACCTGGCTACCAGGACCTGGCCCTGCACTCTCCTGTTTTTTCTTCTCTTCATCCCTGTCTTCTGCAAAG'","metadata":{"trusted":true,"tags":[]},"execution_count":3,"outputs":[],"id":"4b1da235"},{"cell_type":"markdown","source":"Test the third guide! You'll need to turn it into capital letters.","metadata":{},"id":"a43e1890"},{"cell_type":"code","source":"target_of_guide3='gggctctgtttcacctcag'","metadata":{},"execution_count":null,"outputs":[],"id":"458e536e"},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[],"id":"65a0f9fa"},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[],"id":"a0d1e603"},{"cell_type":"markdown","source":"### Lists","metadata":{},"id":"b7054040"},{"cell_type":"markdown","source":"Lists store similar items (strings OR numbers OR logicals). We can use a list to perrform the same operation on some/all of its elements, instead of doing it manually.","metadata":{},"id":"eb7571d6"},{"cell_type":"code","source":"\n# Constructing a list\ncodons = ['TAA', 'tAG', 'ACG','GTA']\nprint(codons)","metadata":{},"execution_count":null,"outputs":[],"id":"859f0507"},{"cell_type":"code","source":"# Accessing an item in a list: use its index -It starts with 0\nfirst_codon = codons[0]\nprint(first_codon)\n","metadata":{},"execution_count":null,"outputs":[],"id":"b9e97485"},{"cell_type":"code","source":"#Print 3th codon","metadata":{},"execution_count":null,"outputs":[],"id":"9978e6a9"},{"cell_type":"code","source":"#What happens if you use a negative index? Try it out","metadata":{},"execution_count":null,"outputs":[],"id":"d563cd1c"},{"cell_type":"code","source":"#What happens if you use an index larger than the list size?","metadata":{},"execution_count":null,"outputs":[],"id":"ae1aa203"},{"cell_type":"code","source":" \n# Modifying an item in a list\ncodons[1] = 'TAG'\nprint(codons)\n","metadata":{},"execution_count":null,"outputs":[],"id":"6227533d"},{"cell_type":"code","source":" \n# Appending an item to the end of a list\ncodons.append('TGA')\nprint(codons)\n \n","metadata":{},"execution_count":null,"outputs":[],"id":"96c582f2"},{"cell_type":"code","source":" \n# Number of items in a list\nnumber_of_codons = len(codons)\nprint('There are', number_of_codons, 'codons')\n","metadata":{},"execution_count":null,"outputs":[],"id":"b7622e45"},{"cell_type":"code","source":" \n# Slicing a list\nsome_codons = codons[2:3]             # \nprint('Second codon:', some_codons)\n","metadata":{},"execution_count":null,"outputs":[],"id":"c360d59f"},{"cell_type":"markdown","source":"### For loop to iterate through a list","metadata":{},"id":"dc6d98cf"},{"cell_type":"code","source":"\n# Save nucleotide bases in a list\nbases = ['T', 'C', 'A', 'G']\n \n# As a warmup,","metadata":{},"execution_count":null,"outputs":[],"id":"476045c5"},{"cell_type":"code","source":"# print the list of bases\nfor base in bases:\n    print(base)","metadata":{},"execution_count":null,"outputs":[],"id":"28f4c19b"},{"cell_type":"code","source":"#Print all codons starting with 'TA'","metadata":{},"execution_count":null,"outputs":[],"id":"dc5dc9e7"},{"cell_type":"code","source":"#Print all codons starting with 'T'","metadata":{},"execution_count":null,"outputs":[],"id":"4727bc74"},{"cell_type":"markdown","source":"### Exercise","metadata":{},"id":"8791da83"},{"cell_type":"markdown","source":"Use a for loop to check which guides have target sequences within the CLA4 gene sequence. This time we have a whole list of guides, so you need to use a FOR loop.","metadata":{},"id":"68fdba80"},{"cell_type":"code","source":"guides = ['CAAGTGAACCTCACTATCCA', \n         'TGGGATCAAAGCTATCTATA',\n         'ATGCCTTCTCCTCAATTCTC',\n         'GATGAGCTGTGGGGAGCCAG',\n         'TGCCTTTGAGGGTCACAACA',\n         'CTGTTTTTTCTTCTCTTCAT',\n         'AATCAATTCTACTAGCTAAT',\n         'CACTATCCAAGGACTGAGGG',\n         'GATTGCCATTTAACATCAAA']","metadata":{"trusted":true,"tags":[]},"execution_count":1,"outputs":[],"id":"cf06bcc4"}]}